### PR TITLE
Added game sounds

### PR DESCRIPTION
**Pull request by husseinalkasake**
**CHANGES**
- Added menu, game and laser sounds (using https://github.com/faiface/beep which seemed to be the most popular)
- Added `go.mod` which might be unnecessary (was added to enforce versioning of packages since the newest version of go-mp3 is needed to avoid crazy issue of memory usage when streaming larger files, there might be a better way to do so) [Record of Mentioned Bug: [here](https://github.com/hajimehoshi/go-mp3/issues/1)]
- Added `audioPlayer.go` in `/resources` which contains all speaker logic stuff (Could argue it belongs in `entities` but I thought it makes sense given it's sole purpose is to play audio resources and doesn't belong on screen)
- Some linting here and there that might be unnecessary / inconsistent

**TODO (Issues to fix)**
- Weird delay when playing files (Tried to debug but couldn't find a way to make it faster. `speaker.Play` method just seems to take a couple of seconds to play which isn't a huge problem for background music but a big one for sound effects). Might have to dig deeper to fix or find another library in the future

### CODE REVIEW

**Comment by alistairfink** [comment]
u think ur sick cuz u formatted ur description?

File: entities/mainMenu.go [hunk 1]
```diff
@@ -54,8 +55,8 @@ func NewMainMenu(resourceLoader *resources.ResourceLoader, win *pixelgl.Window)
...
 }
 
-func (this *mainMenu) Draw() {
-	this.logoSprite.Draw(this.win, this.logoPos)
-	this.menuText.Draw(this.win, this.menuTextPos)
-	this.startText.Draw(this.win, this.startTextPos)
+func (menu *mainMenu) Draw() {
+	menu.logoSprite.Draw(menu.win, menu.logoPos)
+	menu.menuText.Draw(menu.win, menu.menuTextPos)
+	menu.startText.Draw(menu.win, menu.startTextPos)
```
Comment (alistairfink) [inline entities/mainMenu.go]: change these back

File: main.go [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
+	"math/rand"
+	"time"
+
 	"github.com/alistairfink/Hussein-Attack/constants"
 	"github.com/alistairfink/Hussein-Attack/entities"
 	"github.com/alistairfink/Hussein-Attack/resources"
 	"github.com/alistairfink/Hussein-Attack/state"
 	"github.com/faiface/pixel"
 	"github.com/faiface/pixel/pixelgl"
 	"golang.org/x/image/colornames"
```
Comment (alistairfink) [inline main.go]: change these back. they were alphabetized

File: resources/resourceLoader.go [hunk 3]
```diff
@@ -1,10 +1,11 @@
 package resources
 
 import (
-	"github.com/faiface/pixel"
 	"image"
 	_ "image/png"
 	"os"
+
+	"github.com/faiface/pixel"
```
Comment (alistairfink) [inline resources/resourceLoader.go]: Change this back

**Comment by husseinalkasake** [comment]
@alistairfink u don't?

File: resources/audioPlayer.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+import (
+	"log"
+	"os"
+	"time"
+
+	"github.com/faiface/beep"
+	"github.com/faiface/beep/mp3"
+	"github.com/faiface/beep/speaker"
+)
```
Comment (alistairfink) [inline resources/audioPlayer.go:10]: alphabetize these

File: resources/audioPlayer.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		"laser": {
+			index:             2,
+			path:              laserSoundPath,
+			isBackgroundMusic: false,
+		},
+	}
+}
+
+// GetSoundFile returns streamer and format of decoded mp3 file
+func GetSoundFile(soundPath string) (beep.StreamSeekCloser, beep.Format) {
```
Comment (alistairfink) [inline resources/audioPlayer.go]: if this is only used in resources make it private

File: resources/audioPlayer.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+)
+
+type sound struct {
+	index             int
+	path              string
+	isBackgroundMusic bool
+}
+
+// GetSounds gets all sounds for game
+func GetSounds() map[string]*sound {
```
Comment (alistairfink) [inline resources/audioPlayer.go]: should be private if its only used in resources

File: go.mod [hunk 7]
```diff
@@ -0,0 +1,15 @@
+module github.com/alistairfink/Hussein-Attack
+
+go 1.14
```
Comment (alistairfink) [inline go.mod:3]: remove the mod file. this repo doesnt use modules.

File: resources/audioPlayer.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+			defer close(done)
+			select {}
+		} else {
+			// TODO: Make this faster, sound effect buffer doesn't seem to help much
+			speaker.Play(buffer.Streamer(0, buffer.Len()))
+		}
+	}()
+
+	return done
+}
```
Comment (alistairfink) [inline resources/audioPlayer.go:77]: comments are cancer. ur code should be clear enough to read

File: resources/audioPlayer.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+type sound struct {
+	index             int
+	path              string
+	isBackgroundMusic bool
+}
+
+// GetSounds gets all sounds for game
+func GetSounds() map[string]*sound {
+	return map[string]*sound{
```
Comment (alistairfink) [inline resources/audioPlayer.go]: shouldnt these be a global var?

File: resources/audioPlayer.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	return buffer
+}
+
+// PlaySound plays a specific sound (background music or sound effect)
+func PlaySound(context string, buffer *beep.Buffer, initializeSpeaker bool) <-chan bool {
+	done := make(chan bool)
+	go func() {
+		thisSound := GetSounds()[context]
```
Comment (alistairfink) [inline resources/audioPlayer.go]: idk if i like this. seems like bad design to have to refer to each sound as a string. I'd make PlaySound private and have a public function for each sound that just gets the global constant for that sound and passes it into PlaySound. Better coupling?

File: resources/audioPlayer.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+// GetSoundEffectsBuffer returns buffer of all game sound effects to avoid delays/load times
+// (cannot do same for background music as it consumes too much memory)
+func GetSoundEffectsBuffer() *beep.Buffer {
+	// declare buffer
+	var buffer *beep.Buffer
+
+	// loop over and add sound effects to buffer
+	sounds := GetSounds()
+	for k := range sounds {
+		thisSound := sounds[k]
```
Comment (alistairfink) [inline resources/audioPlayer.go]:
```suggestion
	for _, thisSound := range sounds {
```

File: resources/audioPlayer.go (hunk 9 above)
Comment (alistairfink) [reply resources/audioPlayer.go]: also why pointers?

File: resources/audioPlayer.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+	"os"
+	"time"
+
+	"github.com/faiface/beep"
+	"github.com/faiface/beep/mp3"
+	"github.com/faiface/beep/speaker"
+)
+
+type sound struct {
+	index             int
```
Comment (alistairfink) [inline resources/audioPlayer.go]: why index if ur storing in a map?

File: resources/audioPlayer.go [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+	// declare buffer
+	var buffer *beep.Buffer
+
+	// loop over and add sound effects to buffer
+	sounds := GetSounds()
+	for k := range sounds {
+		thisSound := sounds[k]
+		if !thisSound.isBackgroundMusic {
+			streamer, format := GetSoundFile(thisSound.path)
+			if buffer == nil {
```
Comment (alistairfink) [inline resources/audioPlayer.go]: wont this always be null?

**Comment by alistairfink** [comment]
> @alistairfink u don't?

nah. people can read the title if they wanna know wat my pr is

File: go.mod (hunk 7 above)
Comment (alistairfink) [reply go.mod:3]: nvm

File: resources/audioPlayer.go (hunk 13 above)
Comment (husseinalkasake) [reply resources/audioPlayer.go]: currently cuz there's only one sound effect, but in theory if we have a bunch then we should only declare the buffer once

Comment (alistairfink) [reply resources/audioPlayer.go]: oh i see. nvm

File: resources/audioPlayer.go (hunk 8 above)
Comment (husseinalkasake) [reply resources/audioPlayer.go:77]: i removed most, left some that seem worth keeping

File: resources/audioPlayer.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	"os"
+	"time"
+)
+
+type sound struct {
+	path              string
+	isBackgroundMusic bool
+}
+
+var sounds = map[string]sound {
```
Comment (alistairfink) [inline resources/audioPlayer.go]: is there actually a reason we need a map? using global constants for each one would be faster and i think would make more sense than referring to strings

File: resources/filePaths.go [hunk 15]
```diff
@@ -6,3 +6,7 @@ const husseinImageName = "./resources/Hussein.png"
 const laserImageName = "./resources/Laser.png"
 const toiletPaperImageName = "./resources/ToiletPaper.png"
 const virusImageName = "./resources/Virus.png"
+
+const menuMusicPath = "./resources/MenuMusic.mp3"
```
Comment (alistairfink) [inline resources/filePaths.go:12]: The comments that were here were good

Comment (alistairfink) [reply resources/filePaths.go:12]: The logic explaining comments i dont like

File: main.go [hunk 16]
```diff
@@ -46,6 +46,9 @@ func run() {
 	toiletPaperEntities := []entities.ToiletPaper{}
 	virusEntities := []entities.Virus{}
 
+	soundEffectsBuffer := resources.GetSoundEffectsBuffer()
+	resources.PlayMenuMusic()
```
Comment (alistairfink) [inline main.go:50]: Dont change to this but would a singleton instance of a sound manager make more sense for something like this?

File: resources/audioPlayer.go (hunk 14 above)
Comment (alistairfink) [reply resources/audioPlayer.go]: guess it doesnt matter tho since the strings are only referred to in 1 place

**Review by alistairfink — APPROVED** [review]

**Pull request merged**